File: Proyecto/Calendario_App/src/main/java/org/example/calendario_app/dao/impl/UsuarioDAOImpl.java
package org.example.calendario_app.dao.impl;

import org.example.calendario_app.model.Usuario;
import org.example.calendario_app.util.DatabaseConnection;

import java.sql.*;
import java.util.List;

public class UsuarioDAOImpl {
    DatabaseConnection databaseConnection;

    public UsuarioDAOImpl() {
        databaseConnection = new DatabaseConnection();
    }

    public Usuario iniciarSesion(String email) {
        Usuario usuario = null;
        String query = "SELECT * FROM usuarios WHERE correo = ?";
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {
            pstmt.setString(1, email);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    usuario = new Usuario(
                            rs.getInt("id_usuario"),
                            rs.getInt("id_cliente"),
                            rs.getString("correo"),
                            rs.getString("password_hash"),
                            "USER");
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return usuario;
    }

    public int registrar(Usuario usuario) {
        String query = "INSERT INTO usuarios (id_cliente, correo, password_hash) VALUES (?, ?, ?)";
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query, Statement.RETURN_GENERATED_KEYS)) {

            pstmt.setInt(1, usuario.getId_cliente());
            pstmt.setString(2, usuario.getCorreo());
            pstmt.setString(3, usuario.getPassword_hash());

            int affectedRows = pstmt.executeUpdate();
            if (affectedRows > 0) {
                try (ResultSet generatedKeys = pstmt.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        return generatedKeys.getInt(1);
                    }
                }
            }
            return -1;
        } catch (SQLException e) {
            e.printStackTrace();
            return -1;
        }
    }

    public Usuario findByEmail(String email) {
        Usuario usuario = null;
        String query = "SELECT * FROM usuarios WHERE correo = ?";
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {
            pstmt.setString(1, email);
            try (ResultSet rs = pstmt.executeQuery()) {
                if (rs.next()) {
                    usuario = new Usuario(
                            rs.getInt("id_usuario"),
                            rs.getInt("id_cliente"),
                            rs.getString("correo"),
                            rs.getString("password_hash"),
                            "USER");
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return usuario;
    }

    public List<Usuario> findAll() {
        java.util.List<Usuario> usuarios = new java.util.ArrayList<>();
        String query = "SELECT * FROM usuarios";
        try (Connection conn = databaseConnection.getConn();
                Statement stmt = conn.createStatement();
                ResultSet rs = stmt.executeQuery(query)) {
            while (rs.next()) {
                usuarios.add(new Usuario(
                        rs.getInt("id_usuario"),
                        rs.getInt("id_cliente"),
                        rs.getString("correo"),
                        rs.getString("password_hash"),
                        "USER"));
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }
        return usuarios;
    }

    public boolean updatePassword(int userId, String newPasswordHash) {
        String query = "UPDATE usuarios SET password_hash = ? WHERE id_usuario = ?";
        try (Connection conn = databaseConnection.getConn();
                PreparedStatement pstmt = conn.prepareStatement(query)) {
            pstmt.setString(1, newPasswordHash);
            pstmt.setInt(2, userId);
            int rowsUpdated = pstmt.executeUpdate();
            return rowsUpdated > 0;
        } catch (SQLException e) {
            e.printStackTrace();
            return false;
        }
    }
}

File: Proyecto_Android/Calendario_Android_App-31-01-2026/Calendario_Android_App-31-01-2026/Calendario_Android_App/app/src/main/java/com/example/calendario_android_app/dao/impl/UsuarioDAOImpl.kt
package com.example.calendario_android_app.dao.impl

import com.example.calendario_android_app.dao.UsuarioDAO
import com.example.calendario_android_app.model.Usuario
import com.example.calendario_android_app.util.DatabaseConnection
import com.example.calendario_android_app.util.HashUtils
import java.sql.SQLException

class UsuarioDAOImpl : UsuarioDAO {
/**
 * Implementación de la interfaz UsuarioDAO para gestionar el acceso a datos de usuarios.
 * Se comunica con la base de datos MySQL para tareas de autenticación y gestión de perfiles.
 */
class UsuarioDAOImpl : UsuarioDAO {
    
    /**
     * Valida las credenciales de un usuario.
     * @param correo Correo electrónico del usuario.
     * @param contrasena Contraseña en texto plano (será hasheada para comparar).
     * @return Objeto Usuario si las credenciales son válidas, null en caso contrario.
     */
    override fun loguearUsuario(correo: String, contrasena: String): Usuario? {
        val connection = DatabaseConnection.getConnection() ?: return null
        
        var usuario: Usuario? = null
        try {
            // Buscamos al usuario por correo y el hash de la contraseña proporcionada.
            val query = "SELECT * FROM usuarios WHERE correo = ? AND password_hash = ?"
            val statement = connection.prepareStatement(query)
            statement.setString(1, correo)
            statement.setString(2, HashUtils.hashPassword(contrasena))
            
            val resultSet = statement.executeQuery()
            
            if (resultSet.next()) {
                usuario = Usuario(
                    id_usuario = resultSet.getInt("id_usuario"),
                    id_cliente = resultSet.getString("id_cliente"),
                    correo = resultSet.getString("correo"),
                    password_hash = resultSet.getString("password_hash"),
                    rol = resultSet.getString("rol")
                )
            }
            resultSet.close()
            statement.close()
            connection.close()
        } catch (e: SQLException) {
            android.util.Log.e("UsuarioDAOImpl", "Error de SQL en login", e)
            e.printStackTrace()
        }
        return usuario
    }

    /**
     * Comprueba si un correo electrónico ya está registrado en el sistema.
     */
    override fun existeCorreo(correo: String): Boolean {
        val connection = DatabaseConnection.getConnection() ?: return false
        var existe = false
        try {
            val query = "SELECT 1 FROM usuarios WHERE correo = ?"
            val statement = connection.prepareStatement(query)
            statement.setString(1, correo)
            val resultSet = statement.executeQuery()
            if (resultSet.next()) {
                existe = true
            }
            resultSet.close()
            statement.close()
            connection.close()
        } catch (e: SQLException) {
            android.util.Log.e("UsuarioDAOImpl", "Error al comprobar correo", e)
            e.printStackTrace()
        }
        return existe
    }

    /**
     * Registra un nuevo usuario en la base de datos.
     * @param contrasena Se hashea antes de guardarla por seguridad.
     */
    override fun crearUsuario(correo: String, contrasena: String, idCliente: Int): String? {
        val connection = DatabaseConnection.getConnection() ?: return "Error de conexión"
        var errorMessage: String? = null
        try {
            // El rol por defecto es 'usuario'.
            val query = "INSERT INTO usuarios (id_cliente, correo, password_hash, rol) VALUES (?, ?, ?, 'usuario')"
            val statement = connection.prepareStatement(query)
            statement.setInt(1, idCliente)
            statement.setString(2, correo)
            statement.setString(3, HashUtils.hashPassword(contrasena))
            
            val affectedRows = statement.executeUpdate()
            if (affectedRows <= 0) {
                 errorMessage = "No se pudo insertar el usuario"
            }
            statement.close()
            connection.close()
        } catch (e: SQLException) {
            android.util.Log.e("UsuarioDAOImpl", "Error al crear usuario", e)
            e.printStackTrace()
            errorMessage = e.message
        }
        return errorMessage
    }

    /**
     * Obtiene el ID numérico de un usuario a partir de su correo.
     */
    override fun getUsuarioIdByEmail(correo: String): Int? {
        val connection = DatabaseConnection.getConnection() ?: return null
        var id: Int? = null
        try {
            val query = "SELECT id_usuario FROM usuarios WHERE correo = ?"
            val statement = connection.prepareStatement(query)
            statement.setString(1, correo)
            val rs = statement.executeQuery()
            if (rs.next()) {
                id = rs.getInt("id_usuario")
            }
            rs.close()
            statement.close()
            connection.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return id
    }

    /**
     * Busca un usuario por correo electrónico (versión asíncrona suspendida).
     */
    override suspend fun getUserByEmail(correo: String): Usuario? = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var usuario: Usuario? = null
        try {
            connection?.let { conn ->
                val query = "SELECT * FROM usuarios WHERE correo = ?"
                val statement = conn.prepareStatement(query)
                statement.setString(1, correo)
                val rs = statement.executeQuery()
                if (rs.next()) {
                    usuario = Usuario(
                        id_usuario = rs.getInt("id_usuario"),
                        id_cliente = rs.getString("id_cliente"),
                        correo = rs.getString("correo"),
                        password_hash = rs.getString("password_hash"),
                        rol = rs.getString("rol")
                    )
                }
                rs.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        usuario
    }
    
    /**
     * Busca un usuario por su ID único.
     */
    override suspend fun getUserById(idUsuario: Int): Usuario? = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var usuario: Usuario? = null
        try {
            connection?.let { conn ->
                val query = "SELECT * FROM usuarios WHERE id_usuario = ?"
                val statement = conn.prepareStatement(query)
                statement.setInt(1, idUsuario)
                val rs = statement.executeQuery()
                if (rs.next()) {
                    usuario = Usuario(
                        id_usuario = rs.getInt("id_usuario"),
                        id_cliente = rs.getString("id_cliente"),
                        correo = rs.getString("correo"),
                        password_hash = rs.getString("password_hash"),
                        rol = rs.getString("rol")
                    )
                }
                rs.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        usuario
    }

    /**
     * Obtiene información básica (Correo y Nombre) de un usuario.
     * Útil para el envío de notificaciones personalizadas.
     */
    override suspend fun getUserInfo(idUsuario: Int): Pair<String, String>? = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var userInfo: Pair<String, String>? = null
        try {
            connection?.let { conn ->
                val query = """
                    SELECT u.correo, c.nombre 
                    FROM usuarios u
                    LEFT JOIN clientes c ON u.id_cliente = c.id_cliente
                    WHERE u.id_usuario = ?
                """
                val statement = conn.prepareStatement(query)
                statement.setInt(1, idUsuario)
                val rs = statement.executeQuery()
                if (rs.next()) {
                    val email = rs.getString("correo")
                    val name = rs.getString("nombre") ?: "Usuario"
                    userInfo = Pair(email, name)
                }
                rs.close()
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        userInfo
    }
    
    /**
     * Actualiza la contraseña de un usuario en el sistema.
     * La nueva contraseña se hashea antes de ser guardada.
     */
    override suspend fun updatePassword(idUsuario: Int, newPassword: String): Boolean = kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.IO) {
        val connection = DatabaseConnection.getConnection()
        var updated = false
        try {
            connection?.let { conn ->
                val query = "UPDATE usuarios SET password_hash = ? WHERE id_usuario = ?"
                val statement = conn.prepareStatement(query)
                statement.setString(1, HashUtils.hashPassword(newPassword))
                statement.setInt(2, idUsuario)
                val affectedRows = statement.executeUpdate()
                updated = affectedRows > 0
                statement.close()
                conn.close()
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        updated
    }
}
